sdks/python/apache_beam/io/azure/blobstorageio.py

```diff
@@ -0,0 +1,664 @@
+
+    Returns:
+      Dictionary of file name -> size.
+    """
+    storage_account, container, blob = parse_azfs_path(
+        path, blob_optional=True, get_account=True)
+    file_sizes = {}
+    counter = 0
+    start_time = time.time()
+
+    logging.info("Starting the size estimation of the input")
+    container_client = self.client.get_container_client(container)
+
+    while True:
+      response = container_client.list_blobs(name_starts_with=blob)
+      for item in response:
+        file_name = "azfs://%s/%s/%s" % (storage_account, container, item.name)
+        file_sizes[file_name] = item.size
+        counter += 1
+        if counter % 10000 == 0:
+          logging.info("Finished computing size of: %s files", len(file_sizes))
+      break
+
+    logging.info(
+        "Finished listing %s files in %s seconds.",
+        counter,
+        time.time() - start_time)
+    return file_sizes
+
+
+class BlobStorageDownloader(Downloader):
+  def __init__(self, client, path, buffer_size):
+    self._client = client
+    self._path = path
+    self._container, self._blob = parse_azfs_path(path)
+    self._buffer_size = buffer_size
+
+    self._blob_to_download = self._client.get_blob_client(
+        self._container, self._blob)
+
+    try:
+      properties = self._get_object_properties()
+    except ResourceNotFoundError as http_error:
+      if http_error.status_code == 404:
+        raise IOError(errno.ENOENT, 'Not found: %s' % self._path)
+      else:
+        _LOGGER.error(
+            'HTTP error while requesting file %s: %s', self._path, http_error)
+        raise
+
+    self._size = properties.size
+
+  @retry.with_exponential_backoff(
+      retry_filter=retry.retry_on_beam_io_error_filter)
+  def _get_object_properties(self):
+    return self._blob_to_download.get_blob_properties()
+
+  @property
+  def size(self):
+    return self._size
+
+  def get_range(self, start, end):
+    # Download_blob first parameter is offset and second is length (exclusive).
+    blob_data = self._blob_to_download.download_blob(start, end - start)
+    # Returns the content as bytes.
+    return blob_data.readall()
+
+
+class BlobStorageUploader(Uploader):
+  def __init__(self, client, path, mime_type='application/octet-stream'):
+    self._client = client
+    self._path = path
+    self._container, self._blob = parse_azfs_path(path)
+    self._content_settings = ContentSettings(mime_type)
+
+    self._blob_to_upload = self._client.get_blob_client(
+        self._container, self._blob)
+
+    # Temporary file.
+    self._temporary_file = tempfile.NamedTemporaryFile()
+
+  def put(self, data):
+    self._temporary_file.write(data.tobytes())
+
+  def finish(self):
+    self._temporary_file.seek(0)
+    # The temporary file is deleted immediately after the operation.
+    with open(self._temporary_file.name, "rb") as f:
+      self._blob_to_upload.upload_blob(
+          f.read(), overwrite=True, content_settings=self._content_settings)
```

Comment:
I recall an issue related to very large files. What happens when we're trying to upload a large file?

Reply:
@pabloem Let's see:

- Authentication. At the moment the only way to authenticate is with a connection string obtained from environment variables. 
- **Integration tests with Azurite.** Integration tests with Azurite are practically ready. The only thing left is to define a function in `build.gradle` that runs and stops Azurite. (You can find the branch here: [https://github.com/AldairCoronel/beam/commits/azurite](url)).

Reply:
@pabloem What happens when we're trying to upload a large file?

Azure complains when you try to upload large files although in the documentation it states: ***Calls to write a blob, write a block, or write a page are permitted 10 minutes per megabyte to complete. If an operation is taking longer than 10 minutes per megabyte on average, it will time out.***
 
Refer to this issue as well: [https://github.com/Azure/azure-sdk-for-python/issues/12166](url)

Reply:
> Authentication. At the moment the only way to authenticate is with a connection string obtained from environment variables.

Okay this is not acceptable. We need to enable authentication via pipeline options as we already discussed privately. This PR is ready to go, but we need to enable pipelineoptions-based authentication in a follow up, okay?

Also, please address comments by @epicfaace to catch `PartialBatchErrorException`, and then we can move forward to merge this change.